sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala

```diff
@@ -196,14 +194,8 @@ private[hive] object HiveShim {
       if (instance != null) {
         instance.asInstanceOf[UDFType]
       } else {
-        val func = Utils.getContextOrSparkClassLoader
+        Utils.getContextOrSparkClassLoader
           .loadClass(functionClassName).newInstance.asInstanceOf[UDFType]
-        if (!func.isInstanceOf[UDF]) {
```

Comment:
I guess the problem was that caching of the function instance was causing things to break? If so, is the right fix to disable caching for _all_ things which aren't UDFs? Does this risk performance regressions or correctness issues?

I'm not familiar enough with this code to know for sure, so it would be great if you could comment here to help clear this up.

Also, if it does turn out that removing this caching is the correct approach, then I think that we may be able to simply remove `instance`, since it will always be `null` now as far as I can tell.